src/Groups/matrices/forms.jl

```diff
@@ -0,0 +1,447 @@
+#
+# Show
+#
+########################################################################
+
+
+function _assign_description(sym::Symbol)
+   if sym== :alternating print("Alternating")
+   elseif sym== :hermitian print("Hermitian")
+   elseif sym== :symmetric print("Symmetric")
+   elseif sym== :quadratic print("Quadratic")
+   else error("unsupported description")
+   end
+end
+
+
+function Base.show(io::IO, f::SesquilinearForm)
+   _assign_description(f.descr)
+   println(" form with Gram matrix ")
+   show(io, "text/plain", gram_matrix(f))
+end
+
+
+
+
+########################################################################
+#
+# Basic
+#
+########################################################################
+
+#TODO: checking whether two quadratic forms coincide by checking their polynomials is not possible yet.
+==(B::SesquilinearForm, C::SesquilinearForm) = gram_matrix(B)==gram_matrix(C) && B.descr==C.descr
+
+function base_ring(B::SesquilinearForm)
+   if isdefined(B,:matrix) return base_ring(gram_matrix(B))
+   else return base_ring(B.pol)
+   end
+end
+
+"""
+    corresponding_bilinear_form(Q::SesquilinearForm)
+Given a quadratic form `Q`, return the bilinear form `B` defined by `B(u,v) = Q(u+v)-Q(u)-Q(v)`.
+"""
+function corresponding_bilinear_form(B::SesquilinearForm)
+   B.descr==:quadratic || throw(ArgumentError("The form must be a quadratic form"))
+   M = gram_matrix(B)+transpose(gram_matrix(B))
+   if characteristic(base_ring(B))==2 return alternating_form(M)
+   else return symmetric_form(M)
+   end
+end
+
+"""
+    corresponding_quadratic_form(Q::SesquilinearForm)
+Given a symmetric form `f`, returns the quadratic form `Q` defined by `Q(v) = f(v,v)/2`. It is defined only in odd characteristic.
+"""
+function corresponding_quadratic_form(B::SesquilinearForm)
+   B.descr==:symmetric || throw(ArgumentError("The form must be a symmetric form"))
+   characteristic(base_ring(B))!=2 || throw(ArgumentError("Corresponding quadratic form not uniquely determined"))
+   M = deepcopy(gram_matrix(B))
+   l = inv(base_ring(B)(2))
+   for i in 1:nrows(M)
+      for j in i+1:nrows(M)
+         M[j,i]=0
+      end
+      M[i,i]*=l
+   end
+   return quadratic_form(M)
+end
+
+
+
+########################################################################
+#
+# Fields of the variable
+#
+########################################################################
+
+"""
+    gram_matrix(B::SesquilinearForm)
+Return the Gram matrix of a sesquilinear or quadratic form `B`.
+"""
+function gram_matrix(f::SesquilinearForm)
+   isdefined(f,:matrix) && return f.matrix
+
+   @assert f.descr==:quadratic && isdefined(f,:pol) "Cannot determine Gram matrix"
+   d = nvars(parent(f.pol))
+   B = zero_matrix( base_ring(f.pol), d, d )
+   V = collect(exponent_vectors(f.pol))
+   C = collect(coeffs(f.pol))
+   for i in 1:length(V)
+      for j in 1:d
+         if V[i][j] !=0
+            global x = j
```

Comment:
Huh, why does this override a global variable?! That seems wrong

Reply:
But by making it global, you are also potentially overriding the user's variables! That's a big no-go!

If you need the value of `d` outside the loop, simply define x and y beforehand

Reply:
I.e., do something like `x = y = 0` before the first loop

Reply:
Alternatively, use `findfirst`, e.g. `x = findfirst(!iszero, V[i])` (untested, please check that it really does the right thing)

Reply:
Corrected it.